### PR TITLE
Apollo koa integration

### PR DESCRIPTION
**Pull request by HriBB**
[KoaJS](https://github.com/koajs) apollo-server integration. All tests pass, but I still need to test it in a real-life application.

TODO:

- [x] Update CHANGELOG.md with your change (include reference to issue & this PR)
- [ ] Make sure all of the significant new logic is covered by tests
- [ ] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass

### CODE REVIEW

**Comment by helfer** [comment]
@hribb First of all, thanks a lot for the PR!

It looks like all the tests pass, but there are still some linter errors that you should fix:

```
src/integrations/expressApollo.test.ts[28, 26]: block is empty
src/integrations/hapiApollo.test.ts[31, 26]: block is empty
src/integrations/koaApollo.test.ts[16, 10]: Unused import: 'OperationStore'
src/integrations/koaApollo.ts[2, 13]: Unused import: 'koaBody'
```

I'll add comments about the code inline.

File: src/integrations/integrations.test.ts [hunk 1]
```diff
@@ -80,6 +84,7 @@ export default (createApp: CreateAppFunc) => {
                   query: 'query test{ testString }',
               });
           return req.then((res) => {
+              destroyApp(app);
```
Comment (helfer) [inline src/integrations/integrations.test.ts]: Did `afterEach` not work? That would be preferable to adding the call explicitly in every test, I think.

File: src/integrations/koaApollo.test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  app.use(router.allowedMethods());
+  return app.listen(3000);
+}
+
+function destroyApp(app) {
+  app.close();
+}
+
+describe('graphqlHTTP', () => {
+   it('returns express middleware', () => {
```
Comment (helfer) [inline src/integrations/koaApollo.test.ts]: I think this test isn't necessary any more. We could also remove it from the express tests.

File: src/integrations/koaApollo.test.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      });
+      assert.typeOf(middleware, 'function');
+  });
+  it('throws error if called without schema', () => {
+     expect(() => apolloKoa(undefined as ApolloOptions)).to.throw('Apollo Server requires options.');
+  });
+});
+
+describe('renderGraphiQL', () => {
+  it('returns express middleware', () => {
```
Comment (helfer) [inline src/integrations/koaApollo.test.ts]: Like the previous one, this is also not necessary. But if we keep it, it should be renamed.

File: typings.json [hunk 4]
```diff
@@ -6,10 +6,13 @@
   },
   "globalDependencies": {
     "body-parser": "registry:dt/body-parser#0.0.0+20160619023215",
+    "cookies": "registry:dt/cookies#0.5.1+20160316171810",
```
Comment (helfer) [inline typings.json]: Where do we use this? I can't find any reference in the code. Is this something that Koa requires?

File: src/integrations/koaApollo.test.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  expect,
+} from 'chai';
+
+// tslint:disable-next-line
+const request = require('supertest-as-promised');
+
+// tslint:disable-next-line
+import * as koa from 'koa';
+import * as koaRouter from 'koa-router';
+import * as koaBody from 'koa-bodyparser';
```
Comment (helfer) [inline src/integrations/koaApollo.test.ts]: This import isn't used according to tslint.

File: src/integrations/koaApollo.test.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import ApolloOptions from './apolloOptions';
+import { apolloKoa, graphiqlKoa } from './koaApollo';
+
+import { OperationStore } from '../modules/operationStore';
```
Comment (helfer) [inline src/integrations/koaApollo.test.ts]: This import isn't used.

File: src/integrations/koaApollo.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
+import * as koa from 'koa';
+import * as koaBody from 'koa-bodyparser';
```
Comment (helfer) [inline src/integrations/koaApollo.ts]: Unused import according to tslint.

File: src/integrations/expressApollo.test.ts [hunk 8]
```diff
@@ -25,6 +25,8 @@ function createApp(options: CreateAppOptions = {}) {
   return app;
 }
 
+function destroyApp(app) {}
```
Comment (helfer) [inline src/integrations/expressApollo.test.ts]: I think you should get rid of this and instead make the `destroyApp` argument optional. That way you only have to modify one existing test file.

**Comment by HriBB** [comment]
> is this a good plan to introduce the alpha version of koa? will this alienate people who are using 1.x version of koa?

You are right, it's not a good plan, but I am not sure how to support both versions. Problem is that v2 uses a different syntax with async/await, where v1 uses generators/yield.

koa@1.x
```
app.use(function *(){
  this.body = 'Hello World';
});
```
koa@2.x
```
app.use(function(ctx){
  ctx.body = 'Hello World';
});
```
Not really sure how to tackle this. We could simply export two different versions such as `apolloKoa` and `apolloKoa2`. Not the best solution, but it could work, if there is a way to install both versions at once with npm. Is there?

File: typings.json (hunk 4 above)
Comment (HriBB) [inline typings.json]: I guess this is needed by koa typings. If I remove it, I get this error
```
import * as cookies from "cookies";

typings/globals/koa/index.d.ts(5,30): error TS2307: Cannot find module 'cookies'.
```
I am a typescript beginner so I might be doing something wrong. @nnance can you please check this?

**Comment by HriBB** [comment]
@helfer and @nnance I removed all unnecessary tests, fixed linter issues, made `destroyApp()` optional and moved it into `afterEach()` mocha hook. Please check it out and let me know what you think.

**Comment by helfer** [comment]
@HriBB thanks, that looks great! 

Regarding the use of Koa 2 vs. 1, I found this interesting Medium post: https://medium.com/@l1ambda/why-you-should-use-koa-with-node-js-7c231a8174fa#.8h6opnbvy

The tldr is that 2.0 alpha is considered stable, but they won't officially recommend it until Node has native support for async/await and Koa 2.0 can be used without Babel.

I'd say we land this PR as it is, and consider rewriting it for Koa 1 if there are many requests for it.

**Comment by helfer** [comment]
@HriBB can you update the Changelog to mention Koa support? The format should be something along the lines of

* added Koa integration (@HriBB in #59)

**Comment by HriBB** [comment]
@helfer done!

**Comment by HriBB** [comment]
I hope I did not break anything with merging upstream ...

**Comment by helfer** [comment]
Perfect, merged!

**Pull request merged**